tests/test_scraamp.py

```diff
@@ -741,3 +715,211 @@ def test_prescraamp_A_B_join_larger_window(T_A, T_B):
+            for percentage in percentages:
+                seed = np.random.randint(100000)
+
+                np.random.seed(seed)
+                ref_P, ref_I, ref_left_I, ref_right_I = naive.scraamp(
+                    T_B, m, T_B, percentage, zone, False, None, p=p, k=k
+                )
+
+                np.random.seed(seed)
+                approx = scraamp(
+                    T_B,
+                    m,
+                    ignore_trivial=True,
+                    percentage=percentage,
+                    pre_scraamp=False,
+                    p=p,
+                    k=k,
+                )
+                approx.update()
+                comp_P = approx.P_
+                comp_I = approx.I_
+                comp_left_I = approx.left_I_
+                comp_right_I = approx.right_I_
+
+                naive.replace_inf(ref_P)
+                naive.replace_inf(comp_P)
+                npt.assert_almost_equal(ref_P, comp_P)
+                npt.assert_almost_equal(ref_I, comp_I)
+                npt.assert_almost_equal(ref_left_I, comp_left_I)
+                npt.assert_almost_equal(ref_right_I, comp_right_I)
+
+
+@pytest.mark.parametrize("T_A, T_B", test_data)
+@pytest.mark.parametrize("percentages", percentages)
+def test_scraamp_A_B_join_KNN(T_A, T_B, percentages):
+    m = 3
+    for k in range(2, 4):
+        for p in [1.0, 2.0, 3.0]:
+            for percentage in percentages:
+                seed = np.random.randint(100000)
+
+                np.random.seed(seed)
+                ref_P, ref_I, ref_left_I, ref_right_I = naive.scraamp(
+                    T_A, m, T_B, percentage, None, False, None, p=p, k=k
+                )
+
+                np.random.seed(seed)
+                approx = scraamp(
+                    T_A,
+                    m,
+                    T_B,
+                    ignore_trivial=False,
+                    percentage=percentage,
+                    pre_scraamp=False,
+                    p=p,
+                    k=k,
+                )
+                approx.update()
+                comp_P = approx.P_
+                comp_I = approx.I_
+                comp_left_I = approx.left_I_
+                comp_right_I = approx.right_I_
+
+                naive.replace_inf(ref_P)
+                naive.replace_inf(comp_P)
+
+                npt.assert_almost_equal(ref_P, comp_P)
+                npt.assert_almost_equal(ref_I, comp_I)
+                npt.assert_almost_equal(ref_left_I, comp_left_I)
+                npt.assert_almost_equal(ref_right_I, comp_right_I)
+
+
+@pytest.mark.parametrize("T_A, T_B", test_data)
+@pytest.mark.parametrize("percentages", percentages)
+def test_scraamp_plus_plus_self_join_KNN(T_A, T_B, percentages):
+    m = 3
+    zone = int(np.ceil(m / 4))
+    for k in range(2, 4):
+        for p in [1.0, 2.0, 3.0]:
+            for s in range(1, zone + 1):
+                for percentage in percentages:
+                    seed = np.random.randint(100000)
```

Comment:
We have nested 4 for-loops here! Note that when `m==3`, `zone=1` and thus we can remove the `for s in range(1, zone+1)`.  The downside is that we are making the code less flexible (I mean.....if we change `m` later, it does not affect `s`).

Reply:
Obviously, I don't love this. However, if this unit test doesn't take long to complete then it might not matter at all. So let's be pragmatic as nested for-loops only really matter when there are many, many thousands+ things to iterate over in each loop level. In this case, there are probably 10's - 100's of variations, which might not be that bad from a time perspective. Let's start with compute time and then decide if we need to do anything.

I would feel very differently if this were in the actual codebase (and not the unit tests)!

Reply:
> I would feel very differently if this were in the actual codebase (and not the unit tests)!

Yeah...I know :)

---

So, I measured the computing time for different values of `k` (note that we are iterating over  `k in range(2, 4)`). And, I got this:

```
k=1 --> 0.95  
k=2 --> 1.30
k=3 --> 1.61
```
Recall that `k=1` is default, and thus this case is already covered by one of the test function in branch `main`. The new test function covers `k=2` and `k=3`. So, I think it is okay to keep the `for-loop` over `k` as its computing time is small.

Reply:
I agree! Let's keep it